crates/biome_service/src/file_handlers/mod.rs

```diff
@@ -977,7 +978,7 @@ struct LintVisitor<'a, 'b> {
     skip: Option<&'b [RuleSelector]>,
     settings: Option<&'b Settings>,
     path: Option<&'b Utf8Path>,
-    project_layout: Arc<ProjectLayout>,
+    package_json: Option<PackageJson>,
```

Comment:
Shouldn't the visitor still have access to `ProjectLayout`? Why can't we make it so the package json is cached in `ProjectLayout` instead?

Reply:
The `LintParams` still contain have a reference to the `ProjectLayout`, so rules themselves can still access it. Is that what you meant?

Caching *inside* the `ProjectLayout` is something we could still consider if we have further issues with the speed of lookups, but `ProjectLayout` is intended to be easily invalidated, so we would need to be careful which actions invalidate the cache in that case.

Reply:
@dyc3 The PR description was not technically correct, `ProjectLayout::get_node_manifest_for_path` doesn't read a file from the disk. It finds an entry from a cached `HashMap`, matching the path. As [the screenshot](https://github.com/biomejs/biome/pull/5380#issuecomment-2729952393) above, the path matching part is relatively slow and we need to avoid invoking it many times.

I do think the visitor can still have `ProjectLayout`, but it's not used for now except reading the `package.json` for domains.

Reply:
Ah, yes that's what I was concerned with @arendjr. Makes a lot more sense now.